File: 0547-number-of-provinces/0547-number-of-provinces.cpp
class Solution {
public:
    void dfs(vector<vector<int>>& isConnected, int n, int node, vector<int>& visited) {
        visited[node] = 1;
        for (int i = 0; i < n; i++) {
            if (isConnected[i][node] == 1 && !visited[i]) {
                dfs(isConnected, n, i, visited);
            }
        }
    }
    
    int findCircleNum(vector<vector<int>>& isConnected) {
        int provinces = 0;
        int n = isConnected.size();
        vector<int> visited(n, 0);
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                provinces++;
                dfs(isConnected, n, i, visited);
            }
        }
        return provinces;
    }
};

File: 0547-number-of-provinces/0547-number-of-provinces.java
class Solution {
    public int findCircleNum(int[][] isConnected) {
        int n = isConnected.length;
        int numProvinces = 0;
        int[] visited = new int[n];
        for (int i = 0; i < n; i++) {
            if (visited[i] == 0) {
                dfs(isConnected, visited, i, n);
                numProvinces++;
            }
        }
        return numProvinces;
    }
    
    void dfs(int[][] isConnected, int[] visited, int node, int n) {
        visited[node] = 1;
        for (int i = 0; i < n; i++) {
            if (i != node && isConnected[node][i] == 1 && visited[i] == 0) {
                dfs(isConnected, visited, i, n);
            }
        }
    }
}